Invalid "convert the array to a slice" suggestion

### Code

```Rust
trait Test {}
impl Test for &[u8] {}

fn needs_test<T: Test>() -> T {
    panic!()
}

fn example() {
    needs_test::<[u8; 1]>();
    let x: [u8; 1] = needs_test();
}
```


### Current output

```Shell
error[E0277]: the trait bound `[u8; 1]: Test` is not satisfied
 --> src/lib.rs:9:18
  |
9 |     needs_test::<[u8; 1]>();
  |                  ^^^^^^^ the trait `Test` is not implemented for `[u8; 1]`
  |
  = help: the trait `Test` is implemented for `&[u8]`
note: required by a bound in `needs_test`
 --> src/lib.rs:4:18
  |
4 | fn needs_test<T: Test>() -> T {
  |                  ^^^^ required by this bound in `needs_test`
help: convert the array to a `&[u8]` slice instead
  |
9 |     needs_test::<&[u8; 1][..]>();
  |                  +       ++++

error[E0277]: the trait bound `[u8; 1]: Test` is not satisfied
  --> src/lib.rs:10:22
   |
10 |     let x: [u8; 1] = needs_test();
   |                      ^^^^^^^^^^ the trait `Test` is not implemented for `[u8; 1]`
   |
   = help: the trait `Test` is implemented for `&[u8]`
note: required by a bound in `needs_test`
  --> src/lib.rs:4:18
   |
4  | fn needs_test<T: Test>() -> T {
   |                  ^^^^ required by this bound in `needs_test`
help: convert the array to a `&[u8]` slice instead
   |
10 |     let x: [u8; 1] = &needs_test[..]();
   |                      +          ++++

For more information about this error, try `rustc --explain E0277`.
```


### Desired output

```Shell
error[E0277]: the trait bound `[u8; 1]: Test` is not satisfied
 --> src/lib.rs:9:18
  |
9 |     needs_test::<[u8; 1]>();
  |                  ^^^^^^^ the trait `Test` is not implemented for `[u8; 1]`
  |
  = help: the trait `Test` is implemented for `&[u8]`
note: required by a bound in `needs_test`
 --> src/lib.rs:4:18
  |
4 | fn needs_test<T: Test>() -> T {
  |                  ^^^^ required by this bound in `needs_test`

error[E0277]: the trait bound `[u8; 1]: Test` is not satisfied
  --> src/lib.rs:10:22
   |
10 |     let x: [u8; 1] = needs_test();
   |                      ^^^^^^^^^^ the trait `Test` is not implemented for `[u8; 1]`
   |
   = help: the trait `Test` is implemented for `&[u8]`
note: required by a bound in `needs_test`
  --> src/lib.rs:4:18
   |
4  | fn needs_test<T: Test>() -> T {
   |                  ^^^^ required by this bound in `needs_test`

For more information about this error, try `rustc --explain E0277`.
```


### Rationale and extra context

Both of these suggestions result in invalid code, and don't really make much sense. This suggestion should only be emitted when it's about an _expression_ of type `[T; N]`, and not a trait bound that arises some other way (like the type being explicitly written in the code).

### Other cases

_No response_

### Anything else?

This invalid suggestion also occurs in serde autoderived implementations of Deserialize, if a field of the struct/enum is an array with more than 32 elements. (This is the context where I initially discovered this, the MVE above is a reverse-engineering of that case).

This seems to only happen in nightly. I think this is just because this suggestion was added after 1.68.0 was released, or something like that.

The commit that introduced this suggestion (and also the bug) is 8ac7d0eef5621338d3650ec6884c58bc28a5c5f6.

Playground link: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=ec4aeb169fe3a41cfecf995c8cd66150

--- a/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs
+++ b/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs
@@ -3034,6 +3034,7 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
 
             self.maybe_suggest_convert_to_slice(
                 err,
+                obligation,
                 trait_ref,
                 impl_candidates.as_slice(),
                 span,

--- a/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs
+++ b/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs
@@ -3032,7 +3032,7 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                 self.report_similar_impl_candidates_for_root_obligation(&obligation, *trait_predicate, body_def_id, err);
             }
 
-            self.maybe_suggest_convert_to_slice(
+            self.suggest_convert_to_slice(
                 err,
                 obligation,
                 trait_ref,

--- a/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
+++ b/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
@@ -398,7 +398,7 @@ pub trait TypeErrCtxtExt<'tcx> {
         param_env: ty::ParamEnv<'tcx>,
     ) -> Vec<Option<(Span, (DefId, Ty<'tcx>))>>;
 
-    fn maybe_suggest_convert_to_slice(
+    fn suggest_convert_to_slice(
         &self,
         err: &mut Diagnostic,
         obligation: &PredicateObligation<'tcx>,
@@ -3955,7 +3955,7 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
     /// If the type that failed selection is an array or a reference to an array,
     /// but the trait is implemented for slices, suggest that the user converts
     /// the array into a slice.
-    fn maybe_suggest_convert_to_slice(
+    fn suggest_convert_to_slice(
         &self,
         err: &mut Diagnostic,
         obligation: &PredicateObligation<'tcx>,

--- a/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
+++ b/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
@@ -401,6 +401,7 @@ pub trait TypeErrCtxtExt<'tcx> {
     fn maybe_suggest_convert_to_slice(
         &self,
         err: &mut Diagnostic,
+        obligation: &PredicateObligation<'tcx>,
         trait_ref: ty::PolyTraitRef<'tcx>,
         candidate_impls: &[ImplCandidate<'tcx>],
         span: Span,
@@ -3957,10 +3958,17 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
     fn maybe_suggest_convert_to_slice(
         &self,
         err: &mut Diagnostic,
+        obligation: &PredicateObligation<'tcx>,
         trait_ref: ty::PolyTraitRef<'tcx>,
         candidate_impls: &[ImplCandidate<'tcx>],
         span: Span,
     ) {
+        // We can only suggest the slice coersion for function arguments since the suggestion
+        // would make no sense in turbofish or call
+        let ObligationCauseCode::FunctionArgumentObligation { .. } = obligation.cause.code() else {
+            return;
+        };
+
         // Three cases where we can make a suggestion:
         // 1. `[T; _]` (array of T)
         // 2. `&[T; _]` (reference to array of T)

--- a/tests/ui/dst/issue-90528-unsizing-not-suggestion-110063.rs
+++ b/tests/ui/dst/issue-90528-unsizing-not-suggestion-110063.rs
@@ -1,0 +1,13 @@
+trait Test {}
+impl Test for &[u8] {}
+
+fn needs_test<T: Test>() -> T {
+    panic!()
+}
+
+fn main() {
+    needs_test::<[u8; 1]>();
+    //~^ ERROR the trait bound
+    let x: [u8; 1] = needs_test();
+    //~^ ERROR the trait bound
+}

--- a/tests/ui/dst/issue-90528-unsizing-not-suggestion-110063.stderr
+++ b/tests/ui/dst/issue-90528-unsizing-not-suggestion-110063.stderr
@@ -1,0 +1,29 @@
+error[E0277]: the trait bound `[u8; 1]: Test` is not satisfied
+  --> $DIR/issue-90528-unsizing-not-suggestion-110063.rs:9:18
+   |
+LL |     needs_test::<[u8; 1]>();
+   |                  ^^^^^^^ the trait `Test` is not implemented for `[u8; 1]`
+   |
+   = help: the trait `Test` is implemented for `&[u8]`
+note: required by a bound in `needs_test`
+  --> $DIR/issue-90528-unsizing-not-suggestion-110063.rs:4:18
+   |
+LL | fn needs_test<T: Test>() -> T {
+   |                  ^^^^ required by this bound in `needs_test`
+
+error[E0277]: the trait bound `[u8; 1]: Test` is not satisfied
+  --> $DIR/issue-90528-unsizing-not-suggestion-110063.rs:11:22
+   |
+LL |     let x: [u8; 1] = needs_test();
+   |                      ^^^^^^^^^^ the trait `Test` is not implemented for `[u8; 1]`
+   |
+   = help: the trait `Test` is implemented for `&[u8]`
+note: required by a bound in `needs_test`
+  --> $DIR/issue-90528-unsizing-not-suggestion-110063.rs:4:18
+   |
+LL | fn needs_test<T: Test>() -> T {
+   |                  ^^^^ required by this bound in `needs_test`
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0277`.

